editor/src/dev/dev.clj

```diff
@@ -754,3 +773,266 @@
+  (let [primitive-type-kw (buffers/primitive-type-kw component-data-type)
+        read-buf (-> buf
+                     (.slice buf-vertex-attribute-offset attribute-byte-size)
+                     (.order (.order buf))
+                     (buffers/as-typed-buffer component-data-type))]
+    (loop [clj-vector (vector-of primitive-type-kw)]
+      (if (.hasRemaining read-buf)
+        (let [attribute-component (.get read-buf) ; Return type differs by Buffer subclass.
+              clj-vector (conj clj-vector attribute-component)]
+          (recur clj-vector))
+        clj-vector))))
+
+(set! *warn-on-reflection* true)
+
+(defn- buf-clj-vertex-data [^ByteBuffer buf vertex-attributes ^long vertex-stride]
+  (let [buf-size (.limit buf)]
+    (assert (zero? (rem buf-size vertex-stride)) "Buffer size does not confirm to vertex stride.")
+    (mapv (fn [^long buf-vertex-offset]
+            (persistent!
+              (val
+                (reduce (fn [[^long buf-vertex-attribute-offset clj-vertex] vertex-attribute]
+                          (let [attribute-key (:name-key vertex-attribute)
+                                attribute-byte-size (vtx/attribute-size vertex-attribute)
+                                component-data-type (:type vertex-attribute)
+                                clj-vertex-attribute-data (buf-clj-attribute-data buf buf-vertex-attribute-offset attribute-byte-size component-data-type)
+                                clj-vertex-attribute (pair attribute-key clj-vertex-attribute-data)
+                                clj-vertex (conj! clj-vertex clj-vertex-attribute)
+                                buf-vertex-attribute-offset (+ buf-vertex-attribute-offset attribute-byte-size)]
+                            (pair buf-vertex-attribute-offset clj-vertex)))
+                        (pair buf-vertex-offset (transient []))
+                        vertex-attributes))))
+          (range 0 buf-size vertex-stride))))
+
+(defn vertex-buffer-to-clj
+  [^VertexBuffer vbuf]
+  (let [{:keys [attributes ^long size]} (.vertex-description vbuf)
+        vertex-count (count vbuf)
+        buf (.buf vbuf)]
+    {:bytes (* vertex-count size)
+     :count vertex-count
+     :stride size
+     :verts (buf-clj-vertex-data buf attributes size)}))
+
+(defn vertex-buffer-print-data [^VertexBuffer vbuf]
+  (-> (vertex-buffer-to-clj vbuf)
+      (update :verts (fn [verts]
+                       (conj (subvec verts 0 (min 3 (count verts)))
+                             '...)))))
+
+(defmulti matrix-row (fn [matrix ^long _row-index] (class matrix)))
+
+(defmethod matrix-row Matrix3d
+  [^Matrix3d matrix ^long row-index]
+  (let [row (double-array 3)]
+    (.getRow matrix row-index row)
+    row))
+
+(defmethod matrix-row Matrix4d
+  [^Matrix4d matrix ^long row-index]
+  (let [row (double-array 4)]
+    (.getRow matrix row-index row)
+    row))
```

Comment:
I wanted this to be local to the `dev` module, so I didn't introduce a protocol.